### ISSUE TITLE
Problem with asyn device support

### ISSUE THREAD

**Issue by MarkRivers** [epics-modules/busy#5]
The current master branch appears to have an issue.  If callbacks happen quickly it appears to be calling the driver when it should not.  This is showing up in NDPluginFile with AutoSave=1.  There is a [new issue for this in ADCore](https://github.com/areaDetector/ADCore/issues/291). 

ADCore has currently worked around the issue.  

It needs to be tracked down and fixed in busy, but I don't have time for it this week.

busy R1-7 should not be released until this problem is fixed.

**Comment by MarkRivers** [comment]
This is a more detailed description of the issue. The problem is that devBusyAsyn.c does not behave properly when there are rapid callbacks from the driver. When there are callbacks from the driver this is what should happen:

- The record should process once for each callback value
- The device support should detect that it is being called due to a driver callback, and should not call the driver with this value from the callback.

The version of devBusyAsyn.c starting with 2c1cfca916767651b92d90bc8ef81f07b1212b79 fails to do both of these things.

- It only had a single variable to store the callback values from the driver, not a ring buffer. So it even when it processed multiple times it was always using the last callback value, not all callback values.
- It did have a counter to keep track of how many callbacks had occurred since the record last processed. However, it did not decrement this counter when it processed, it set it to zero and printed a warning message if it was greater than 1.
- Because each driver callback did result in a call to scanOnce(), the record would process N times if there were N callbacks in a row before it processed. But only the first one was properly handled and did not result in a call to the driver. The remaining N-1 times it would call the driver.

There was an additional problem with the logic. It did not correctly distinguish between processing due to driver callbacks, and processing due to a real write to the record. If the record is actually written to this value must be sent to the driver, but this was not guaranteed if the driver was doing rapid callbacks.

This last problem is shared with the standard asyn device support for output records that have the info tag asyn:READBACK. They also do not have a way to distinguish between processing due to a record callback and that due to a write to the record, and will require a similar fix to that needed for devBusyAsyn.c

**Comment by MarkRivers** [comment]
This issue should be fixed with pull request #6 which does the following:

- The driver callback function (interruptCallback) now allocates a structure containing the callback value from the driver, a CALLBACK structure, and a pointer to the record.  
- It calls callbackRequest using the standard EPICS callback tasks, with the driverProcessCallback function to be called.
- driverProcessCallback does the following:
  - Locks the record
  - Sets the flag saying the record is being processed due to a driver callback
  - Copies the callback value into the device private structure
  - Processes the record
  - Unlocks the record
  - Frees the structure allocated above

This should fix all of the problems.  
- Each callback request is passed a different structure containing the callback value for that callback.  
- When device support is called it knows if this is due to a driver callback or not.   It only write the value to the driver if it is not processing due to a driver callback.

A potential problem with this solution is that there is no protection against driver callbacks happening so fast that the record cannot keep up.  If this happens then the callback task queue will fill up and errors will be generated.  One way to fix this would be to use a ring buffer and not make the callback request if the ring buffer is full.  This would be a blend of how asyn device support for output records currently works and this new callback method.  asyn currently senses that it should get the value from the ring buffer if the ring buffer is not empty.  The hybrid method would use the ring buffer to hold the values, but a flag set in the callback function to indicate why the record is processing.

**Comment by MarkRivers** [comment]
I have come up with a different solution.

I changed asyn/devEpics/devAsynInt32.c to fix the problem with callback processing on output records.  The problem was that it was not distinguishing processing due to driver callback (where the driver should not be called) and normal record processing (where the driver should be called).  I have fixed that by calling callbackRequest() rather than scanOnce() in the interrupt callback function for output records.  The callback function ensures that the device support knows why the record is processing and behaves correctly.

I then copied that new version of devAsynInt32.c into devBusyAsyn.c, and basically just deleted unneeded code and replaced "boRecord" with "busyRecord".  It appears to work OK.

Note that I first tried adding the busyRecord device support to devAsynInt32.c in asyn.  This would reverse the dependencies of asyn and busy.  Rather than busy being optionally dependent on asyn, asyn would be optionally dependent on busy.  This is a "cleaner" solution in that only a small amount of code needs to be added to devAsynInt32.c to support the busy record, while devBusyAsyn.c is basically just a lot of duplicated code from devAsynInt32.c.  However, reversing the dependencies requires changes to a large number of application dbd files, since busySupport.dbd now needs to come before asyn.dbd, rather than after it.  In the interest of backwards compatibility I decided the redundant code was the lesser of the 2 evils.
